flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSink.java

```diff
@@ -31,14 +31,24 @@
 
 	private static final long serialVersionUID = 1L;
 
+	private transient SimpleSinkContext sinkContext;
+
 	public StreamSink(SinkFunction<IN> sinkFunction) {
 		super(sinkFunction);
 		chainingStrategy = ChainingStrategy.ALWAYS;
 	}
 
 	@Override
+	public void open() throws Exception {
+		super.open();
+
+		this.sinkContext = new SimpleSinkContext<>();
+	}
+
+	@Override
 	public void processElement(StreamRecord<IN> element) throws Exception {
-		userFunction.invoke(element.getValue());
+		sinkContext.element = element;
+		userFunction.invoke(sinkContext, element.getValue());
```

Comment:
wouldn't it be better/simpler to just pass `StreamRecord` to the `userFunction`?

`userFunction.invoke(element)`?

and instead of adding `SinkContext` as a first argument of the `invoke` method in the sink interface, just change the element type from `IN` to `StreamRecord<IN>`?

Reply:
`StreamRecord` is an internal concept that should not be exposed in any user facing API. That's the reason for the extract context. Plus, the context allows us to extend the information that we pass to the `SinkFunction` in the future, similar to the context in `ProcessFunction` and `ProcessWindowFunction`.

Reply:
I get it, however I still do not like this objects repacking thing done here. Instead of passing `StreamRecord` directly, we could introduce `UserRecord` interface with `getTimestamp()` method, that would be implemented by `StreamRecord`.

But that's minor complain, if you have stronger preferences for keeping it as it is, it's fine for me.